libs/content/workspace/src/space_inspector/show.rs

```diff
@@ -0,0 +1,494 @@
+    }
+
+    pub fn reset_root(&mut self) {
+        self.data.focused_folder = self.data.root;
+        self.paint_order = vec![];
+    }
+
+    pub fn get_color(&self, curr_id: Uuid, mut layer: usize, mut child_number: usize) -> Color32 {
+        let big_table = [
+            // red
+            [
+                Color32::from_rgb(128, 15, 47),
+                Color32::from_rgb(164, 19, 60),
+                Color32::from_rgb(201, 24, 74),
+                Color32::from_rgb(255, 77, 109),
+                Color32::from_rgb(255, 117, 143),
+                Color32::from_rgb(255, 143, 163),
+            ],
+            // green
+            [
+                Color32::from_rgb(27, 67, 50),
+                Color32::from_rgb(45, 106, 79),
+                Color32::from_rgb(64, 145, 108),
+                Color32::from_rgb(82, 183, 136),
+                Color32::from_rgb(116, 198, 157),
+                Color32::from_rgb(116, 198, 157),
+            ],
+            // blue
+            [
+                Color32::from_rgb(2, 62, 138),
+                Color32::from_rgb(0, 119, 182),
+                Color32::from_rgb(0, 150, 199),
+                Color32::from_rgb(0, 180, 216),
+                Color32::from_rgb(72, 202, 228),
+                Color32::from_rgb(144, 224, 239),
+            ],
+        ];
+        if layer == 1 {
+            if child_number > 2 {
+                child_number %= 3;
+            }
+            return big_table[child_number][0];
+        }
+
+        let parent_color = self
+            .colors
+            .iter()
+            .find(|item| item.id == self.data.all_files.get(&curr_id).unwrap().file.parent)
+            .unwrap()
+            .color;
+
+        let parent_type = big_table
+            .iter()
+            .enumerate()
+            .find_map(|(row_index, row)| {
+                row.iter()
+                    .position(|&x| x == parent_color)
+                    .map(|col_index| (row_index, col_index))
+            })
+            .unwrap()
+            .0;
+
+        layer -= 1;
+
+        if layer > 5 {
+            layer %= 6;
+        }
+
+        big_table[parent_type][layer]
+    }
+
+    pub fn follow_paint_order(
+        &mut self, ui: &mut Ui, root_anchor: Rect, window: Rect,
+    ) -> Option<Uuid> {
+        let mut changed_focused_folder: Option<Uuid> = None;
+        let mut current_layer = 0;
+        let mut current_position = root_anchor.min.x;
+        let mut child_number = 1;
+        let mut visited_folders: Vec<DrawHelper> = vec![];
+        let mut current_parent =
+            DrawHelper { id: self.data.focused_folder, starting_position: 0.0 };
+
+        for (i, item) in self.paint_order.iter().enumerate() {
+            let item_filerow = self.data.all_files.get(&item.id).unwrap();
+
+            if current_layer != item.layer {
+                current_position = root_anchor.min.x;
+                current_layer = item.layer;
+            }
```

Comment:
I was surprised that this function is not recursive while the other ones are and when I read these lines I have a feeling things are about to get unnecessarily complicated. In the other recursive functions, it's clear when you're moving up a layer (recursive call) or down a layer (return), but now you're forced to infer those things from the layer info. It's hard to be sure at this point that all the mutable local variables that needed to be updated have been updated.

Reply:
I happened to spend a bunch of time in this code while playing with colors so I do get what's going on here, but I remember it taking a while, and I remember creating a bug anytime I tried to change anything (that was an old version tho)

Reply:
(this feedback not necessarily actionable - I'm not going to ask you to rework a thing that's working but I hope you read this code in a year and feel like you would have an easier time)